traits/trait_list_object.py

```diff
@@ -53,460 +88,750 @@ def __init__(self, index=0, removed=None, added=None):
+    # TraitList interface
+    # ------------------------------------------------------------------------
 
-        # Do the validated 'setslice' assignment without raising an
-        # 'items_changed' event:
-        if trait.minlen <= len(value) <= trait.maxlen:
-            try:
-                validate = trait.item_trait.handler.validate
-                if validate is not None:
-                    value = [validate(object, name, val) for val in value]
+    def validate(self, index, removed, value):
+        """ Validate values for given index and removed values.
 
-                list.__setitem__(self, slice(0, 0), value)
+        This simply calls the validator provided by the class, if any.
+        The validator is expected to have the signature::
 
-                return
+            validator(trait_list, index, removed, value)
 
-            except TraitError as excp:
-                excp.set_prefix("Each element of the")
-                raise excp
+        and return a list of validated values or raise TraitError.
 
-        self.len_error(len(value))
+        Parameters
+        ----------
+        index : int or slice
+            The indices being modified by the operation.
+        removed : object or list
+            The item or items to be removed.
+        value : object or iterable
+            The new item or items being added to the list.
 
-    def _send_trait_items_event(self, name, event, items_event=None):
-        """ Send a TraitListEvent to the owning object if there is one.
+        Returns
+        -------
+        values : object or list
+            The validated values
+
+        Raises
+        ------
+        TraitError
+            If validation fails.
         """
-        object = self.object()
-        if object is not None:
-            if items_event is None and hasattr(self, "trait"):
-                items_event = self.trait.items_event()
-            object.trait_items_event(name, event, items_event)
+        # Use getattr as pickle can call `extend` before validator is set.
+        if getattr(self, 'validator', None) is None:
+            if isinstance(index, slice):
+                return list(value)
+            else:
+                return value
+        else:
+            return self.validator(self, index, removed, value)
+
+    def notify(self, index, removed, added):
+        """ Call all notifiers
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(trait_list, index, removed, added)
+
+        Any return values are ignored.
+
+        Parameters
+        ----------
+        index : int or slice
+            The indices being modified by the operation.
+        removed : object or list
+            The item or items to be removed.
+        added : object or list
+            The new item or items being added to the list.
```

Comment:
Could we normalize this interface such that `removed` and `added` are always a list, i.e. use a list of one item to replace a single object?  Otherwise downstream code is going to have to do isinstance check to differentiate one item versus many items. This could potentially make handling of list of list difficult?

Reply:
I agree, `Undefined` has been removed and `added` and `removed` are now lists.